src/ucp/dt/datatype_iter.inl

```diff
@@ -104,24 +104,78 @@ static UCS_F_ALWAYS_INLINE ucs_status_t ucp_datatype_iter_iov_allocate_memh(
     return UCS_OK;
 }
 
+static int
+ucp_datatype_iter_is_user_memh_valid(const ucp_datatype_iter_t *dt_iter,
+                                     const ucp_mem_h memh)
+{
+    ucs_string_buffer_t err_msg = UCS_STRING_BUFFER_INITIALIZER;
+    int is_valid                = 0;
+    int memh_contains_buffer;
+    size_t iov_count;
+
+    if (memh == NULL) {
+        ucs_error("got NULL memory handle");
+        goto out;
+    }
+
+    if (memh->mem_type != dt_iter->mem_info.type) {
+        ucs_error("unmatched memory type [%s]"
+                  " memh %p [%s]",
+                  ucs_memory_type_names[dt_iter->mem_info.type], memh,
+                  ucs_memory_type_names[memh->mem_type]);
+        goto out;
+    }
+
+    switch (dt_iter->dt_class) {
+    case UCP_DATATYPE_CONTIG:
+        memh_contains_buffer = ucp_memh_is_buffer_in_range(
+                memh, dt_iter->type.contig.buffer, dt_iter->length, &err_msg);
+        break;
+    case UCP_DATATYPE_IOV:
+        iov_count            = ucp_datatype_iter_iov_count(dt_iter);
+        memh_contains_buffer = ucp_memh_is_iov_buffer_in_range(
+                memh, dt_iter->type.iov.iov, iov_count, &err_msg);
+        break;
+    default:
+        ucs_error("unsupported memory handle datatype: [%s]",
+                  ucp_datatype_class_names[dt_iter->dt_class]);
+        goto out;
+    }
+
+    if (!memh_contains_buffer) {
+        ucs_error("mismatched memory handle [%s]"
```

Comment:
1. here says "mismatched", in line 122 says "unmatched"
2. print what is mismatched (the buffer)

Reply:
number 2 is already done in `ucp_memh_is_buffer_in_range`

Reply:
so we would have 2 error messages printed?
need to consolidate

Reply:
It will be one message, but it's  built in 2 steps: first the buffer details and then memh details

Reply:
so what will be the complete error message in this case?

Reply:
AFAIR I tried to stick with the same format which was until now:
for contig:
     mismatched memory handle [buffer 0x123 length 123] memh 0x234 [address 0x345 length 345]
for IOV:
     mismatched memory handle [buffer 0x123 length 123 iov[2]] memh 0x234 [address 0x345 length 345]

Reply:
How about:
`mismatched memory handle 0x234 [address 0x345 length 345] for [buffer 0x123 length 123]`
and IOV:
`mismatched memory handle 0x234 [address 0x345 length 345] for iov[1] [buffer 0x123 length 123]`